### PR TITLE
8320258: Refresh building.md

### PR DESCRIPTION
**Pull request by magicus**
The contents of the build README has been poorly kept up to date in places. With the move to Github, the need to have markdown syntax that looks good on Github has become apparent. The entire document has been in need for a while of a comprehensive oversight to be made more consistent in language and formatting.

Here is a summary of the most important changes:

* The entire section on cross-compiling have been restructured to be more logical, and some parts have been rewritten, so it reads more coherent.
* A section on "make doctor" has been added
* The fixpath.sh script has been explained more in detail (and references to the old fixpath binary removed)
* Information about MSYS2 and WSL has been updated
* References to how things were done in the past, like the hg forest have been removed
* Version numbers have been updated where needed
* Some specific paragraphs about e.g. boot JDK detection have been clarified
* Example output have been updated
* Remaining references to `run-test` have been updated to just `test`
* Internal links have been fixed in a few places
* A few remaining places that confuses "OpenJDK" with the "JDK" have been fixed
* Markdown formatting has been improved in some places
* The entire file has gotten more normalized markdown formatting (fix bullet list indentation, add empty lines surrounding the code block markers, etc)

Note that this PR will can possibly create a merge conflict with 8317357 and 8264425; my intention is to integrate these two first and then resolve any conflicts in this PR.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320258](https://bugs.openjdk.org/browse/JDK-8320258): Refresh building.md (**Bug** - P3)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16696/head:pull/16696` \
`$ git checkout pull/16696`

Update a local copy of the PR: \
`$ git checkout pull/16696` \
`$ git pull https://git.openjdk.org/jdk.git pull/16696/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16696`

View PR using the GUI difftool: \
`$ git pr show -t 16696`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16696.diff">https://git.openjdk.org/jdk/pull/16696.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16696#issuecomment-1814944483)

### CODE REVIEW

**Comment by magicus** [comment]
The resulting markdown file can be seen here: https://github.com/magicus/jdk/blob/readme-refresh/doc/building.md
The resulting HTML can be seen here: https://htmlpreview.github.io/?https://github.com/magicus/jdk/blob/readme-refresh/doc/building.html

File: doc/building.md [hunk 1]
```diff
@@ -1261,59 +1334,61 @@ Architectures that are known to successfully cross-compile like this are:
 | sh4          | sid          | sh4           | sh4-linux-gnu            | zero                      |
 | riscv64      | sid          | riscv64       | riscv64-linux-gnu        | (all)                     |
 
-### Building for ARM/aarch64
+### Considerations for specific targets
+
+#### Building for ARM/aarch64
```
Comment (erikj79) [inline doc/building.md:1339]: I don't think we need to or recommend setting ABI for aarch64.

File: doc/building.md [hunk 2]
```diff
@@ -1947,9 +2064,52 @@ In the JDK build, we always use Unix paths internally, and only just before
 calling a tool that does not understand Unix paths do we convert them to
 Windows paths.
 
-This conversion is done by the `fixpath` tool, which is a small wrapper that
-modifies unix-style paths to Windows-style paths in command lines. Fixpath is
-compiled automatically by `configure`.
+This conversion is done by the `fixpath.sh` tool, which is a small wrapper that
+modifies Unix-style paths to Windows-style paths. The fixpath
+tool is called with the first argument as a sub-command describing the action
```
Comment (erikj79) [inline doc/building.md:2069]: Looks like the text needs to be reflowed here.

File: doc/building.md [hunk 3]
```diff
@@ -1947,9 +2064,52 @@ In the JDK build, we always use Unix paths internally, and only just before
...
+
+* `verify` is called at configure time to check that a path is correctly
+  specified and reachable by Windows.
+
+The fixpath tool uses a somewhat complex heuristic to infer which part of the
+command line arguments refer to paths, and converts those. In some
+circumstances, these heuristics can fail.
+
+If you are having strange build issues related to path conversion, you might
+need to debug how fixpath treats your paths. There are many ways to do this.
```
Comment (erikj79) [inline doc/building.md:2090]: many -> "several", or maybe "multiple"? Or perhaps "Here are some ways to do this."

**Review by erikj79 — COMMENTED** [review]
Thank you for taking on the complete file, this looks much better. Just a few comments.

File: doc/building.md [hunk 4]
```diff
@@ -39,18 +39,26 @@ technical expertise, a fair number of dependencies on external software, and
 reasonably powerful hardware.
 
 If you just want to use the JDK and not build it yourself, this document is not
-for you. See for instance [OpenJDK installation](
-http://openjdk.org/install) for some methods of installing a prebuilt
-JDK.
+for you. See for instance [OpenJDK installation](http://openjdk.org/install)
```
Comment (vidmik) [inline doc/building.md:42]: Switch to https? Same for other uses of http:// later in the file?

Comment (erikj79) [reply doc/building.md:42]: There is a separate PR posted for that specific change already.

File: doc/building.md [hunk 5]
```diff
@@ -238,13 +246,38 @@ please check the Cygwin FAQ on the ["BLODA" list](
...
+
 #### Windows Subsystem for Linux (WSL)
 
-Windows 10 1809 or newer is supported due to a dependency on the wslpath utility
-and support for environment variable sharing through WSLENV. Version 1803 can
-work but intermittent build failures have been observed.
+WSL comes in two flavors, WSL1 and WSL2. These are drastically different under
+the hood. WSL1 runs the binaries natively by translating Linux kernel calls
+into Windows kernel calls, while WSL2 runs Linux in a virtual machine. Both
+solutions has their pros and cons, and you might need to test both before
```
[the remark this answers is not in the document]
Comment (magicus) [reply doc/building.md:273]: My numerus nemesis strikes again. :-/ Thanks for noticing.

File: doc/building.md [hunk 6]
```diff
@@ -3,11 +3,11 @@
 ## TL;DR (Instructions for the Impatient)
 
 If you are eager to try out building the JDK, these simple steps works most of
```
Comment (RealCLanger) [inline doc/building.md:5]: could change works->work here

File: doc/building.md (hunk 1 above)
Comment (magicus) [reply doc/building.md:1339]: Yep. I'll fix the heading.

**Comment by magicus** [comment]
I have now checked the grammar of the entire document with automated tools. :) I pushed some of the suggested changes.
